controller/src/controller/models/domain/rsptag.py

```diff
@@ -576,3 +578,86 @@ def subset(
 
         # Return the results.
         return type(self)(tags)
+
+    def by_type(self, image_type: RSPImageType) -> Self:
+        """Return all tags of a given type.
+
+        Parameters
+        ----------
+        image_type
+            RSPImageType to select from collection.
+
+        Returns
+        -------
+        RSPImageCollection
+            Collection of tags matching the given type.
+        """
+        return type(self)(self._by_type[image_type])
+
+    def apply_policy(self, policy: RSPImageFilterPolicy) -> Self:
+        """Apply a filter policy and return the remaining tags.
+
+        Note that this relies on the ordering of types, as subset() does.
+
+        Parameters
+        ----------
+        policy
+            Policy governing tag filtering.
+
+        Returns
+        -------
+        RSPImageTagCollection
+            Tags remaining after policy application.
+        """
+        tags: list[RSPImageTag] = []
+        for category in RSPImageType:
+            if category == RSPImageType.ALIAS:
+                # All alias tags are preserved.
+                tags.extend(list(self.by_type(category).all_tags()))
+                continue
+            if category == RSPImageType.UNKNOWN:
+                # No unknown tags are preserved.
+                continue
+            tags.extend(
+                self._apply_category_policy(
+                    policy,
+                    category,
+                )
+            )
+        return type(self)(tags)
+
+    def _apply_category_policy(
+        self,
+        policy: RSPImageFilterPolicy,
+        category: RSPImageType,
+    ) -> list[RSPImageTag]:
+        candidates = list(self.by_type(category).all_tags())
+        remainder: list[RSPImageTag] = []
+
+        # Mypy needs to take it on faith that this is how you map
+        # category names to attributes of the overall policy,
+        # each one being an ImageFilterPolicy.
+        pol_attr = category.value.lower().replace(" ", "_")
+        cat_policy = cast(ImageFilterPolicy, getattr(policy, pol_attr))
+
+        for tag in candidates:
+            # First check the count; if we're at our limit, we're done
+            if cat_policy.number is not None and cat_policy.number <= len(
+                remainder
+            ):
+                break
+            # Apply age policy
+            tag_age = tag.age()
+            if tag_age is not None and cat_policy.age is not None:
+                if tag_age > cat_policy.age:
+                    continue
+            # Apply version policy
+            if (
+                tag.version is not None
+                and cat_policy.cutoff_version is not None
+            ):
+                if tag.version < cat_policy.cutoff_version:
+                    continue
```

Comment:
Looks like you don't need the `age` method on an `RSPImageTag` at all (or the `by_type` method above). This method does need to take an argument giving the basis for the age calculation so that it's consistent for all calculations. I'm calling that `age_basis` here.

```suggestion
        cutoff_version = policy.cutoff_version
        cutoff_date = None
        if policy.age:
            cutoff_date = age_basis - policy.age
        for tag in self._by_type[category]:
            if policy.number is not None and len(remainder) == policy.number:
                break
            if cutoff_version is not None:
                if tag.version is not None and tag.version < cutoff_version:
                    continue
            if cutoff_date and tag.date and tag.date < cutoff_date:
                continue
```

This will get more of the code on one line and should make it easier to follow. The comments are just repeating the code, so I'm not sure they add much.

Reply:
Fair, although I prefer doing the cutoff_date and version-to-semver calculations inside the "do we have a policy for this kind of filter, and do we have enough info on the tag to apply it" bit, when we're already committed to applying that filter.